transport/ice/IceSession.cpp

```diff
@@ -472,11 +493,252 @@ bool IceSession::isResponseAuthentic(const void* data, size_t len) const
         stunMessage->isAuthentic(_hmacComputer.remote);
 }
 
+bool IceSession::isIceAuthentic(const void* data, size_t len) const
+{
+    if (ice::isResponse(data))
+    {
+        return isResponseAuthentic(data, len);
+    }
+
+    if (ice::isRequest(data))
+    {
+        return isRequestAuthentic(data, len);
+    }
+
+    return false;
+}
+
+bool IceSession::processValidStunTcpRequest(IceEndpoint* localEndpoint,
+    const transport::SocketAddress& remotePort,
+    const StunMessage& msg,
+    int remoteCandidatePriority,
+    uint64_t now)
+{
+    assert(localEndpoint->getTransportType() == TransportType::TCP);
+
+    if (isAttached(localEndpoint))
+    {
+        sendResponse(localEndpoint, remotePort, 0, msg, now);
+        return true;
+    }
+
+    IceCandidate remoteCandidate(_component,
+        localEndpoint->getTransportType(),
+        remoteCandidatePriority,
+        remotePort,
+        remotePort,
+        IceCandidate::Type::PRFLX);
+
+    auto* candidatePair = addRemoteTcpCandidateAndCreatePair(remoteCandidate, localEndpoint);
+    if (candidatePair)
+    {
+        candidatePair->accept();
+        sendResponse(localEndpoint, remotePort, 0, msg, now);
+        candidatePair->receptionTimestamp = now;
+        candidatePair->send(now);
+    }
+
+    if (_state == State::CONNECTING)
+    {
+        sortCheckList();
+    }
+
+    return candidatePair != nullptr;
+}
+
+bool IceSession::processValidStunUdpRequest(IceEndpoint* localEndpoint,
+    const transport::SocketAddress& remotePort,
+    const StunMessage& msg,
+    int remoteCandidatePriority,
+    uint64_t now)
+{
+    assert(localEndpoint->getTransportType() == TransportType::UDP);
+
+    if (!(_state == State::CONNECTING || _state == State::CONNECTED))
+    {
+        return false;
+    }
+
+    auto* pair = findCandidatePair(localEndpoint, remotePort);
+    if (pair)
+    {
+        // Only send response after accept! Otherwise we could drop early media
+        pair->accept();
+        sendResponse(localEndpoint, remotePort, 0, msg, now);
+        return true;
+    }
+
+    const auto* remoteCandidate = findCandidateWithUdpAddress(_remoteCandidates, remotePort);
+    if (remoteCandidate)
+    {
+        // if it is not a new candidate but we don't have a candidate that could mean we are receiving
```

Comment:
I wonder if it is better to avoid this logic. If we discarded the candidate, it should be for a reason. Reviving it could be a problem in case we do want to ignore certain candidates. We should get the pruning logic right instead. Perhaps lopg this but just reject the candidate.

Reply:
I will do that. But as we are always trying to keep `remoteCandidate` short, I guess this is unlikely (if not impossible) to detect. We will see like if it was a new PRFLX candidate